Document Acceptance Criteria for Random Beacon v2

michalinacienciala opened the pull request:
The document stores User Acceptance Criteria for Random Beacon v2
feature. It will be used to gain common understanding of how Random
Beacon v2 should act and will be utilised during the acceptance testing
of the feature.

michalinacienciala committed:
Document Acceptance Criteria for Random Beacon v2

The document stores User Acceptance Criteria for Random Beacon v2
feature. It will be used to gain common understanding of how Random
Beacon v2 should act and will be utilised during the acceptance testing
of the feature.

michalinacienciala committed:
Add/rephrase UACs

Some new UACs has been added to the list. Bunch of UACs has been updated
to better reflect the expected behaviour.

michalinacienciala committed:
Document resetting of DKG timeout timer by challenge notification

It was agreed that:
"In case of a challenge we will get number of blocks that was mined until
the challange happened and use it as an offset to:
1. determine new starting block for submissions, so members start
   submitting from the first member
2. calculate new DKG timeout to give all members a chance to submit
   a result"

tomaszslabon commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+  - notifier should receive 5% from the total slashed amount (in the same tx) +
+  - the members of the signing group should be given another chance to publish
+  the DKG result
+  - DKG timeout timer and the result subission eligibility order shoul be reset.
```
Some typos: 
subission -> submission
shoul -> should

michalinacienciala replied:
Fixed, see 7cc86c2e707b6b13450709e6f596f933b4c2439c.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+  `IRandomBeaconConsumer` interface.
+
+| Relay request - relay entry submission
+| If the transaction submitting the new relay entry gets approved, then: +
```
We should also mention about exceeding soft timeout and bleeding. This is because we perform the actual slashing of the bled out value upon entry submission.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+  blocks for a member to become eligible to submit relay entry. The timer starts
+  at the moment when the first member becomes eligible (the moment when the
+  relay entry was requested). +
+  The hard relay entry timeout is a set value (measured in time units).
```
Bear in mind the hard timeout timer starts at the moment when soft timeout is reached. The default value is `24h` but we use blocks to represent it in code. We assume `1 block = 15s`.

michalinacienciala replied:
Clarified, see 953ccbb65131b597e4fe3bba5ac4dca810aadcf1.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+  soft relay entry timeout) to a fixed value (at the moment of hard relay entry
+  timeout).
+
+| Relay request - relay entry timeouts
+| If no entry was successfully submitted until soft relay timeout is reached,
+  all operators in the group should start bleeding and losing their T stake. The
+  amount of lost stake should incrementally increase from 0 (at the moment of
+  soft relay entry timeout) to a fixed value (at the moment of hard relay entry
+  timeout).
```
Duplication?

michalinacienciala replied:
Right, removed, see 953ccbb65131b597e4fe3bba5ac4dca810aadcf1.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+  - has the hard relay entry timeout passed? (if not, then notification invalid) +
+
+| Relay request - relay entry timeouts
+| If the hard relay entry timeout notification was deemed justified, then: +
```
Also:
- All members of the terminated group should be slashed up to the fixed slashing amount
- If there are no active groups, relay entry could not be repeated. The request should be cleaned up to allow new requests to happen.

michalinacienciala replied:
Added, see 953ccbb65131b597e4fe3bba5ac4dca810aadcf1 and 3cb023632eeaf6e1143d26413657da592d2b99aa.

lukasz-zimnoch reviewed (commented):
The first part, related to the relay entry.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+  `IRandomBeaconConsumer` interface.
+
+| Group creation
+| New groups should be crated with a fixed frequency of relay request (every
```
Typo `crated` -> `created`

michalinacienciala replied:
Fixed, see 7cc86c2e707b6b13450709e6f596f933b4c2439c.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+  generation should be repeated.
+
+| Relay entry utilisation
+| Applications wanting to use a relay entry should submit another transaction
```
Should we allow to easily read the recent relay entry?

michalinacienciala replied:
Hmm, I'm not sure what you mean by easily, but generally I believe that application should be able to retrieve the info about the recent entry. The UAC is taken straight from the [RFC-19](https://github.com/keep-network/keep-core/blame/main/docs/rfc/rfc-19-random-beacon-v2.adoc#L191).

lukasz-zimnoch replied:
Sorry, I wasn't precise enough. I mean should the `RandomBeacon` expose a dedicated view function to check the `previousEntry`? Now the entry is emitted as an event which makes it possible to fetch but a bit problematic for off-chain clients. For on-chain clients, the entry is passed within the callback.

michalinacienciala replied:
Claarified in 6aa27e527b74c6bc8705469207369e9bf0979c7a.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+  then (when the pool gets unlocked) for joining operators to pools.
+
+| Group creation
+| Group creation start transaction should be embedded into relay entry request
```
The group creation should be embedded into the relay entry submission transaction. Looks like there is a mistake in the RFC.

michalinacienciala replied:
@lukasz-zimnoch, so the new groups will be created with some specific frequency of the relay entry *requests*, but the trigger for the creation of group will be included in the relay entry *submission* tx, right?

So let's say the new groups are created every 3 relay requests.
req_x - group G1 submitted the relay entry, no creation of a new group requested
req_x+1 - group G2 submitted the relay entry, no creation of a new group requested
req_x+2 - group G3 submitted the relay entry, requested the creation of a new group
req_x+3 - group G4 submitted the relay entry, no creation of a new group requested
req_x+4 - group G5 submitted the relay entry, no creation of a new group requested
req_x+5 - group G6 failed to submit the relay entry and was terminated, group G7 was selected to create the entry, G7 submitted the relay entry and requested the creation of a new group
(G1-G7 do not need to be different groups)

Do I get that right?

lukasz-zimnoch replied:
Yeah. Each request has a `requestId` starting from `1`. If `requestId mod groupCreationFrequency == 0`, the submission of relay entry corresponding to that request will trigger group creation by emitting the `DkgStarted` event.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+  - triggering of the DKG (distributed key generation) +
+  - the beacon should start the period of DKG result submission eligibility.
+
+| Croup creation - selecting members
```
Typo `Croup` -> `Group`

michalinacienciala replied:
Fixed, see 7cc86c2e707b6b13450709e6f596f933b4c2439c.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,438 @@
+  aspects should be taken into account: +
+  - is the submitter eligible at the moment to send the tx? +
+  - are there enough supporting signatures on the result? +
+  - do the signatures come from the stakers with at least a minimum stake? +
```
They can't join/leave the pool during the lock but their staked amount can still change. For example, they can be slashed due to relay entry timeout. If a member becomes ineligible (their stake drops below the minimum) during a lock, they will be removed from the pool as soon as a status update occurs.

michalinacienciala replied:
Clarified, see e5d9a5989baaf4a860d0007855a1556af7d1781d.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+| When determining validity of the tx submitting the DKG result, the following
+  aspects should be taken into account: +
+  - is the submitter eligible at the moment to send the tx? +
+  - are there enough supporting signatures on the result? +
```
...and all those signatures are valid.

michalinacienciala replied:
Added, see 74353bb0ffc7205f4059cb08a6b0535cafd5e3c6.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,465 @@
+  deemed justified, then: +
+  - the malicious DKG result should be immediately discarded (in the same
+  transaction in which notification happened) +
+  - all sortition pool members who signed the result should be slashed (fixed
+  amount) (in the same tx) +
+  - notifier should receive 5% from the total slashed amount (in the same tx) +
```
IIRC, it should be burned.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+| Group creation - DKG timeout
+| The DKG submission timeout equals the group size multiplied by the number of
+  blocks for a member to become eligible to submit the DKG result. The timer
+  starts at the moment when the first member becomes eligible (the moment when
```
We should include the off-chain DKG time. It's currently `72` blocks.

michalinacienciala replied:
Fixed, see 68f430e5fcb4f81e65aa832ca7c79b8f3774d297. I did not provide the exact value, just added the info that the added time should cover for the time of DKG generation.
Btw, is this `72 blocks` a governable param? If so, I will need to add another UAC for that.

lukasz-zimnoch replied:
Nope, it's not. It's a fixed off-chain duration of the DKG protocol.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,465 @@
+| If the DKG timeout notification was deemed justified, then: +
+  - the pool should be unlocked +
+  - the fixed amount reward (in T) should be sent from the maintenance pool to
+  the notifier.
```
In relay entry, we punish using slashing if the timeout has been violated. However, IIRC it's not the case here. We probably won't punish for DKG timeouts (RFC doesn't mention about that as well) as they are less harmful than relay entry timeouts. @pdyraga is that correct?

pdyraga replied:
Yes, let's not punish them for now. It could happen that some operators dropped from the pool and honest ones could not produce DKG result. Being in the group and not answering calls is a serious crime. Dropping and not cooperating in DKG is less harmful to the network.

michalinacienciala replied:
Ok, thanks for clarification, I don't change anything here then.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+| Sortition pool - punishments
+// TODO: revisit 2 weeks removal time
+| After 2 weeks punish period elapses, operators removed from the sortition pool
+  should be automatically reinstated in the pool and should be able to earn the
```
Currently, they need to re-register manually. Is the `automatically reinstated` a hard requirement?

michalinacienciala replied:
No, this is just how I thought it will work, but there were no requirements regarding this, so I believe the manual re-registration is fine. Changed description in 60895c09c5c4dd137f632135b7170b7cd7783f8e.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+| When the sortition pool is locked, no operator can enter the pool.
+
+| Sortition pool - locking
+| When the ssortition pool is locked, no operator can leave the pool.
+
+| Sortition pool - locking
+| Once the ssortition pool gets unlocked, operators can leave the pool.
```
Typos `ssortition`

michalinacienciala replied:
Fixed, see 7cc86c2e707b6b13450709e6f596f933b4c2439c.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+  - for successful submission of the DKG result (fixed value) +
+  - for successful unlocking of the sortition pool if DKG timed out.
+
+| Sortition pool - locking
```
I think we should also mention that during locking, operators should not be able to update their state. This is because they can be removed during the update.

michalinacienciala replied:
Added, see e5d9a5989baaf4a860d0007855a1556af7d1781d.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+| Once the ssortition pool gets unlocked, operators can leave the pool.
+
+| Sortition pool - locking
+| Once the ssortition pool gets unlocked, new operators can join the pool.
```
Typo `ssortition`

michalinacienciala replied:
Fixed, see 7cc86c2e707b6b13450709e6f596f933b4c2439c.

lukasz-zimnoch reviewed (commented):
An the next part

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,438 @@
+
+| Group creation
+| Group creation start transaction should be embedded into relay entry request
+  transaction
```
See https://github.com/keep-network/keep-core/pull/2665#discussion_r742780503

michalinacienciala replied:
Fixed, see 2da13b067f573385dae3ba9822eee33b9768ec2e.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,438 @@
+  - after fixed amount of blocks another member (with index
+  `(hash(new_group_pubkey) % group_size ) + 1` should become eligible (previous
+  operator remains eligible) +
+  - next members should become eligible after each fixed amount of blocks
```
Yep, this the first operator is random in that case.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,438 @@
+  all operators in the group should start bleeding and losing their T stake. The
+  amount of lost stake should incrementally increase from 0 (at the moment of
+  soft relay entry timeout) to a fixed value (at the moment of hard relay entry
+  timeout)
```
Exactly as you said.

michalinacienciala replied:
Fixed, see 953ccbb65131b597e4fe3bba5ac4dca810aadcf1.

michalinacienciala committed:
Fix typos

michalinacienciala committed:
Document what happens with the tokens slashed due to DKG challenge

michalinacienciala committed:
Clarify what happens when relay entry can't be produced

michalinacienciala committed:
Clarify validation of the fee allowance for a realy request

michalinacienciala committed:
Explain what the maintanance pool is

michalinacienciala committed:
Add a couple of UACs regarding the locking of the sortition pool

michalinacienciala committed:
Failed BLS threshold signature results in hard timeout

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+| Anyone can send a relay entry request transaction.
+
+| Relay request
+| Processing of each relay entry request should always result in the generation
```
Clarified, see 3cb023632eeaf6e1143d26413657da592d2b99aa.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+| When determining validity of the relay entry request, the following aspects
+  should be taken into account: +
+  - isn’t the other relay entry request already pending? +
+  - has the requester provided the exact required fee? (if the fee is lower or
```
Clarified, see 452961fc2dc5faa44736f752a180873ef3decc1c.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+| Relay request
+| If the relay entry request is deemed valid, then: +
+  - the fee provided by the requester should be deposited in the maintenance
+  pool (with an intention to cover the reimbursements for DKG-related actions) +
```
Clarified, see 4264d5835edc2e592363d08e7219b6bc696c8cb3.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,438 @@
+| Once the active group is selected as a signing group, it is tasked with the
+  generation of a BLS threshold signature (based on the current relay entry and
+  the data provided by at least `blsThreshold=(group_size/2) + 1` members of the
+  signing group).
```
Clarified, see a334799625c786e007a45694a56c46d6577074de.

michalinacienciala committed:
Clarify how re-registration of operators in the pool should work

michalinacienciala committed:
Validity of the signatures should be checked during DKG submission

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,438 @@
+| When determining validity of the relay entry submission, the following aspects
+  should be taken into account: +
+  - is the submitter eligible to submit? +
+  - hasn’t the hard timeout for the relay request be reached?
```
@lukasz-zimnoch, where this restriction will be implemented? In the client? If yes, then wouldn't it be possible for some operator to submit the relay entry outside of the client? (I'm not sure if that's possible - just wanna make sure we won't miss some corner case)

lukasz-zimnoch replied:
It's implemented on the contract level.

michalinacienciala committed:
Embed creation of new group into relay entry submission tx

michalinacienciala committed:
Improve description of relay entry timeouts and slashing

Also a duplicate UAC removed.

michalinacienciala committed:
Add validation criteria for relay entry submission

michalinacienciala committed:
Include DKG geneartion time in the DKG timeout counter

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,467 @@
+
+| Relay request - relay entry submission
+| When determining validity of the relay entry submission, the following aspects
+  should be taken into account: +
```
Added, see baeb53b03c55e33d5a226a7ec2d9c491514f5d24.

lukasz-zimnoch reviewed (commented):
Looks very good! No more comments on my side.

michalinacienciala committed:
Remove duplicated UAC

michalinacienciala committed:
Modify method of punishment for misbehaviour

Initially we wanted to exclude the misbehaving operators by removing
them from the sortition pool for a fixed amount of time. But this posed
a problem in the following scenario:
- DKG starts, pool is locked
- Relay entry is requested
- Some members are inactive and they missed their turn to submit the
  result
On the one hand, this would require removing some members from the
sortition pool. On the other, no oerators should leave the pool when
it's locked.
As a solution it was decided that misbehaving operators will not be
removed from the pool, instead they will not be eligible to earn the
rewards for a fixed, governable amount of time.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,438 @@
+  higher, the transaction should be reverted) +
+  - is there at least one active signing group?
+
+| Relay request
```
I updated the UACs based on what was agreed upon: e9cdc51244cf831adde0b177a9d2d3c7b21c35b2.

michalinacienciala replied:
Ok, so that I think is mostly in line whit what I added in that e9cdc51244cf831adde0b177a9d2d3c7b21c35b2 commit, but I missed a couple of places where I should have do the update. Fixed in c090c77d63a13e4caa536ea50f07f71ec435b5d6.

pdyraga commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  pool to the submitter’s address (in the same tx) +
+  - operators who became eligible for submitting the DKG result before the
+  member who submitted the DKG result should be punished by being excluded from
+  earning rewards for a fixed amount of time (in the same tx).
```
There is one additional scenario here not covered by the RFC: we need to have a fixed timeout under which DKG result submitter is expected to call the approve transaction. If they do not do that, anyone becomes eligible to call the approve transaction. The one who called that transaction receives the submitter reward.

This protection is needed because DKG result approve transaction might not be that cheap to call and we want to have everyone incentivized to call it in case DKG result submitter decides not doing that.

michalinacienciala replied:
Clarified who can mark the result as accepted in commit 8c9a2ffd19dd2435957041357aaaabc3aa4d75cd. Should this fixed timeout be a governable param? If yes, I need to add a UAC for it.

michalinacienciala replied:
Atill not sure if that's governable or not. Asked also in https://github.com/keep-network/keep-core/pull/2751/files#r762768804.

pdyraga commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  - misbehavior during the DKG submission +
+  - misbehavior during relay entry submission.
+
+| Sortition pool - punishments
+| Operators punished with removal from the group should not earn rewards for the
+  time of their absence in the pool.
```
We are not removing them from the pool, we are banning them from rewards for a fixed, governable period of time.

michalinacienciala replied:
Right, I missed that, fixed in c090c77d63a13e4caa536ea50f07f71ec435b5d6.

pdyraga commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  time of their absence in the pool.
+
+| Sortition pool - punishments
+| The operation removing members from the sortition pool should be as cheap as
+  possible so that the member submitting the relay entry or the DKG result does
+  not have to pay any additional costs for removing inactive members from the
+  pool.
```
Again, we are not removing them. We are banning them from rewards.

michalinacienciala replied:
Right, I missed that, fixed in c090c77d63a13e4caa536ea50f07f71ec435b5d6.

pdyraga commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,465 @@
+  not have to pay any additional costs for removing inactive members from the
+  pool.
+
+| Sortition pool - rewards
+| Operators in the sortition pool can earn T for multiple activities: +
+  - for staying in the pool (given weekly, proportionally to stake) +
+  - for successful submission of the DKG result (fixed value) +
+  - for successful unlocking of the sortition pool if DKG timed out.
```
I am fine with having it here.

pdyraga commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+// add them to the UACs
+
+| Governable parameters
+| The frequency of relay requests should be a governable parameter.
```
Hmm.... no? Applications using the random beacon decide about it.

michalinacienciala replied:
Right, the frequency applies to creation of new groups. Improved the description in b0ed2dd.

pdyraga commented on docs/rfc/rfc-19-random-beacon-v2.adoc:
```diff
@@ -61,9 +61,9 @@ operators noticed the problem.
-chain, all off-chain clients should perform group selection optimistically, 
+transaction submitting relay entry and should lock a sortition pool. From this
+moment, no operator can enter or leave the pool. Once a new relay entry appears
+on the chain, all off-chain clients should perform group selection optimistically, 
```
Negative. The pool is locked in the relay request transaction. The event to start DKG off-chain is emitted in relay entry transaction. The seed used for the sortition is unknown at the moment the pool is locked.

michalinacienciala replied:
Right, the frequency applies to creation of new groups. Improved the description in b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

lukasz-zimnoch commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  - at the beginning (right after the challenge period passes) only one member
+  with index `hash(new_group_pubkey) % group_size` should be eligible +
+  - the first eligible operator should be given additional time for submitting
+  the result, to cover for the time of off-chain DKG generation
```
Yep, we're not giving additional time for the first submitter.

michalinacienciala replied:
Removed, see 3.
But we still add the additional 72 blocks to DKG timeout counter, as per comment https://github.com/keep-network/keep-core/pull/2665#discussion_r742794248, right?
If so, I'm not sure why we would want to add the blocks there but not to the eligibility window of the first chosen DKG result submitter. Don't we risk a situation that the DKG will last so long that the first potential submitter will not have enough time to publish the result?

michalinacienciala committed:
Improve wording, make descriptions more precise

michalinacienciala committed:
Clarify when group creation starts and what are its stages

michalinacienciala committed:
Don't add the DKG generation time to eligibility window of 1st operator

michalinacienciala committed:
Punish inactive operators with excluding from rewards

michalinacienciala committed:
Lock sortition pool in relay request transaction

michalinacienciala committed:
Improve description of UACs related to DKG result

michalinacienciala committed:
Clarify punishing for relay entry timeout

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+toc::[]
+
+== Purpose of the document
+This document stores User Acceptance Criteria for Random Beacon v2 feature. It
```
Changed, see b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+| UAC
+
+| Relay request
+| Anyone can send a relay entry request transaction.
```
Good point, changed in b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+| Relay request
+| Successful processing of each relay entry request should always result in the
+  generation of a random number (relay entry) and sometimes also (if certain
+  conditions are met) in a creation of a new signing group.
```
Fixed in b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  - the fee charged from the requester should be deposited in the maintenance
+  pool (the T balance of the random beacon contract) (with an intention to cover
+  the reimbursements for DKG-related actions) +
+  - the current relay entry should be used by the beacon to select a signing
```
Fixed, see b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  group from the set of active groups (by hashing the relay entry and applying
+  `modulo no_of_active_groups` operation on the result) +
+  - the random beacon contract should emit an event informing about the signing
+  group selection +
```
Fair point, changed in b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+| If the relay entry request is deemed valid, then: +
+  - the fee charged from the requester should be deposited in the maintenance
+  pool (the T balance of the random beacon contract) (with an intention to cover
+  the reimbursements for DKG-related actions) +
```
Fixed, see b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  - at the beginning (right after the challenge period passes) only one member
+  (with index `new_entry % group_size` should be eligible +
+  - after fixed amount of blocks another member (with index `(new_entry %
+  group_size ) + 1` should become eligible (previous operator remains eligible) +
```
Fixed, see b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+| Relay request - relay entry submission
+| When requesting a relay entry, it should be possible to pass an optional
+  address parameter - this is the address of a contract implementing
+  `IRandomBeaconConsumer` interface.
```
Right, it should, moved in b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+
+| Relay request - relay entry submission
+| The transaction submitting relay entry should have a predictable cost and the
+  gas cost of this transaction should be as low as possible, below 200k gas.
```
Fixed, see b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+| Relay request - relay entry submission
+| The operation removing members from the sortition pool should be as cheap as
+  possible so that the member submitting relay entry does not have to pay any
+  additional costs for removing inactive members from the pool.
```
Right, I missed that, fixed in c090c77d63a13e4caa536ea50f07f71ec435b5d6.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  amount of lost stake should incrementally increase from 0 (at the moment of
+  soft relay entry timeout) to a fixed value (at the moment of hard relay entry
+  timeout). The slashing will be applied in transaction: +
+  - submitting the relay entry submission (if the result got submitted by one of
```
Fixed, see b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  notification invalid) +
+  - has the hard relay entry timeout passed? (if not, then notification invalid) +
+  - does the transaction include punishing of the misbehaving operators? (if not,
+  then notification invalid).
```
By "punishing of the misbehaving operators" I meant punishing of all operators in the group (as they all misbehaved). Rephrased in a67f737fecaa50e14b4a535f13cf9bb493c2600d.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+| If the hard relay entry timeout notification was deemed justified, then: +
+  - the signing group should get terminated +
+  - all signing group members should be punished by being slashed the fixed
+  percentage of T and being excluded from earning rewards for a fixed amount of
```
Fixed, see a67f737fecaa50e14b4a535f13cf9bb493c2600d.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+
+| Group creation
+| Group creation start transaction should lock the sortition pool and emit the
+  group creation start event.
```
Updated, see a65103938718759d4d370c2b7bc3866e1c951f8b.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  group creation start event.
+
+| Group creation
+| Group creation start transaction should be as cheap as possible.
```
Fixed, see b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+| Group creation
+| Group creation start transaction should result in: +
+  - selection of a candidate group consisting of `group_size` members of
+  sortition pool +
```
Thanks for the clarification, I updated the descriptions as part of a65103938718759d4d370c2b7bc3866e1c951f8b.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+| Upon noticing the group creation start event, clients should choose the
+  sortition pool members to the candidate group based on the current relay
+  entry, ensuring that the higher is the stake of an operator, the higher is his
+  chance of being selected to the group.
```
Updated, see a65103938718759d4d370c2b7bc3866e1c951f8b.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+| At a given moment, only selected members from the group should be eligible to
+  successfully submit the DKG result to the chain: +
+  - at the beginning (right after the challenge period passes) only one member
+  with index `hash(new_group_pubkey) % group_size` should be eligible +
```
Sure, that was a typo on my side, corrected in a65103938718759d4d370c2b7bc3866e1c951f8b.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  - within or outside of the challenge period for the specified DKG result (if
+  outside, then invalid) +
+  - have all members who signed a DKG result been selected by the sortition pool
+  to be in the group (if yes, then challenge invalid).
```
Fixed, see 8c9a2ffd19dd2435957041357aaaabc3aa4d75cd.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  - all sortition pool members who signed the result should be slashed (fixed
+  amount) (in the same tx) +
+  - notifier should receive 5% from the total slashed amount (in the same tx) +
+  - remaning 95% of the total slashed amount shoud be burned +
```
Changed, see b0ed2dd71ecdf63263790d0a5bf7670bfb3a1a16.

michalinacienciala committed:
Clarify UACs relating to callbacks

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+
+| Relay entry utilisation
+| Applications wanting to use a relay entry should submit another transaction
+  using the relay entry value previously set by the random beacon.
```
Changed, see 6aa27e527b74c6bc8705469207369e9bf0979c7a.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+  terminated and no result should be produced (in order for the next requests to
+  get processed successfully, the manual genesis must be first triggered).
+
+| Relay entry utilisation
```
Changed, see 6aa27e527b74c6bc8705469207369e9bf0979c7a.

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,503 @@
+
+| Relay request - relay entry submission
+| It should be possible to send low gas callbacks (up to the fixed gas cost
+  value) in the transaction submitting the relay entry.
```
I see, I misunderstood the RFC, thought that the callbacks and calling `IRandomBeaconConsumer` are two different things. Removed in 6aa27e527b74c6bc8705469207369e9bf0979c7a.

michalinacienciala commented:
@Piotr's comments addressed, ready for another round of reviews.

michalinacienciala committed:
Clarify what constitutes a malitious DKG result

michalinacienciala committed:
Give names to governable parameters

michalinacienciala committed:
Clarify who gets reward for accepting DKG result

michalinacienciala committed:
Describe punishing of operators IA/DQ during DKG protocol

michalinacienciala committed:
Document UAC around group expiration

Each group created in the system remains active for a certain period of
time. A group that expired is no longer selected for any new work. Group
expiration is performed in the relay request transaction.

michalinacienciala committed:
Document value of callback gas limit

michalinacienciala committed:
Reward the notifier of hard timeout

michalinacienciala committed:
`maintainance pool` -> `rewards pool`

michalinacienciala committed:
Clarify that 200k gas used limit does not include callback cost

michalinacienciala committed:
Don't verify eligibility to submit relay request on-chain

michalinacienciala committed:
Describe `dkgMaliciousResultNotificationRewardMultiplier`

michalinacienciala committed:
Describe punishing of the operators ignoring their duties

michalinacienciala committed:
Describe punishing for leaked signature

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,607 @@
+| The length of the challenge period (`resultChallengePeriodLength`) should be a
+  governable parameter and should be initally set to `1440 blocks`.
+
+| Governable parameters
+| The slashing amount for submitting malicious DKG result
+  (`maliciousDkgResultSlashingAmount`) should be a governable parameter.
```
Not sure if the slashing is for submitting the result or supporting. Asked in https://github.com/keep-network/keep-core/pull/2751/files#r761879103.

michalinacienciala committed:
Update list of actions leading to the rewards ban

michalinacienciala commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,607 @@
+  aspects should be taken into account: +
+  - does it reference the existing DKG request? (if not, then notification
+  invalid) +
+  - has the DKG timeout passed? (if not, then notification invalid) +
```
Not sure if that condition need to be always passing. 

From the Random beacon README:
> DKG timeout includes the situation when no new relay entry was produced and sortition could not be performed.

That means this situation:
Somebody requested a new entry, frequency of group creation was hit, and due to that the sortition pool got locked. Then nobody published the new relay entry and we now want to unlock the pool and will do that in the DGK timeout notification.

From RFC-1 we know that:
> The [DKG timeout] timer starts at the moment when the first member becomes eligible [to submit DKG result]

My understanding is that in the case I described we never start the DKG, so we can't say whether the DKG timeout has passed or not...

michalinacienciala commented:
UACs list updated after recent changes to the Random Bacon v2. Ready for another round of reviews.

tomaszslabon commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,608 @@
+  - next members should become eligible after each fixed amount of blocks.
+
+| Relay request - punishing inactive operators
+| It should be possible for a member of the group to send a notication informing
```
Just a few typos: 
notication = > notification
elgible, eliglible=> eligible
genearte => generate
remaning => remaining
malitious => malicious
initally => initially 
notouriously => notoriously 
transfered => transferred

tomaszslabon commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,608 @@
+  random beacon should determine its validity.
+  
+| Relay request - punishing inactive operators
+  When determining validity of the transaction notifying about the misbehaving
```
We should probably have `|` before `When determining validity of the transaction notifying about the misbehaving`. It probably causes the `Area` and `UAC` columns mixed up when we view [this document](https://github.com/keep-network/keep-core/blob/random-beacon-uacs/docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc) with a browser.

michalinacienciala replied:
Right, fixed in 2866f5ea524741025d92299c89b1bd2441274b52.

michalinacienciala committed:
Fix grammar & typos

michalinacienciala committed:
Fix table layout

tomaszslabon commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,608 @@
+
+| Governable parameters
+| The length of the challenge period (`resultChallengePeriodLength`) should be a
+  governable parameter and should be initially set to `1440 blocks`.
```
It seems the value of `resultChallengePeriodLength` has been updated recently in [this PR](https://github.com/keep-network/keep-core/pull/2749) and is now `11520` blocks (~48h assuming 15s block time).

tomaszslabon commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,608 @@
+| Governable parameters
+| Every `N`-th relay request initiates a creation of a new signing group.
+  This group creation frequency (`groupCreationFrequency`) should be a governable
+  parameter.
```
We could mention that `5` was chosen as the initial `groupCreationFrequency` in [this PR](https://github.com/keep-network/keep-core/pull/2749).

tomaszslabon commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,608 @@
+| Governable parameters
+| The time since the creation of a new signing group after that group gets
+  expired (`groupLifetime`) should be a governable parameter and should be
+  initially set to `80640 blocks` (~2 weeks, assuming 15s block time).
```
In [this PR](https://github.com/keep-network/keep-core/pull/2749) we updated the initial value for `groupLifetime` to `403200`  ( ~10 weeks assuming 15s block time).

tomaszslabon commented on docs/acceptance-criteria/uacs-rfc-19-random-beacon-v2.adoc:
```diff
@@ -0,0 +1,608 @@
+
+| Governable parameters
+| The slashing amount for submitting malicious DKG result
+  (`maliciousDkgResultSlashingAmount`) should be a governable parameter.
```
The initial value of `maliciousDkgResultSlashingAmount` is `50000 T` on the `main` branch.

michalinacienciala committed:
Update values of governable parameters

Now that we have a set of proposed values for governable parameters in
the RandomBeacon contract, we can update the UACs.

michalinacienciala commented:
@tomaszslabon, I applied your propositions, also updated a couple more of parameters.
